bigquery_etl/shredder/delete.py

```diff
@@ -0,0 +1,337 @@
+        raise exceptions[-1]
+    return results
+
+
+def run_query(client, query, dry_run, priority):
+    """Run a query in bigquery and wait for it to complete."""
+    job = client.query(
+        query, bigquery.QueryJobConfig(dry_run=dry_run, priority=priority)
+    )
+    if not dry_run:
+        job.result()
+    return job
+
+
+def get_partition_expr(table):
+    """Get the SQL expression to use for a table's partitioning field."""
+    for field in table.schema:
+        if field.name != table.time_partitioning.field:
+            continue
+        if field.field_type == "TIMESTAMP":
+            return f"DATE({field.name})"
+        return field.name
+
+
+def list_partitions(client, table):
+    """List the partition ids and corresponding dates in a table."""
+    return [
+        (partition_id, datetime.strptime(partition_id, "%Y%m%d").date())
+        for row in client.query(
+            f"SELECT partition_id FROM [{table.full_table_id}$__PARTITIONS_SUMMARY__]",
+            bigquery.QueryJobConfig(use_legacy_sql=True),
+        ).result()
+        for partition_id in row.values()
+    ]
+
+
+async def delete_from_partition(
+    client_q,
+    executor,
+    dry_run,
+    priority,
+    target,
+    partition_id,
+    clustering_fields,
+    **template_kwargs,
+):
+    """Process deletion requests for a single partition of a target table."""
+    client = client_q.default_client
+    template_kwargs["target"] = target
+    if not target.cluster_conditions:
+        queries = [DELETE_TEMPLATE.format(**template_kwargs)]
+    else:
+        clustering = f"CLUSTER BY {', '.join(clustering_fields)}"
+        expiration_date = datetime.utcnow().date() + timedelta(days=15)
+        queries = [
+            SELECT_TEMPLATE.format(
+                # expire in 15 days because this script may take up to 14 days
+                expiration_timestamp=f"{expiration_date} 00:00:00 UTC",
+                destination=get_temporary_table(client),
+                clustering=(clustering if needs_clustering else ""),
+                cluster_condition=cluster_condition,
+                **template_kwargs,
+            )
+            for cluster_condition, needs_clustering in target.cluster_conditions
+        ]
+    run_tense = "Would run" if dry_run else "Running"
+    for query in queries:
+        logging.debug(f"{run_tense} query: {query}")
+    loop = asyncio.get_running_loop()
+    jobs = await gather_all(
+        loop.run_in_executor(
+            executor, client_q.with_client, run_query, query, dry_run, priority
+        )
+        for query in queries
+    )
+    # copy results into place if necessary
+    if target.cluster_conditions:
+        sources = [
+            f"{t.project}.{t.dataset_id}.{t.table_id}"
+            for t in [job.destination for job in jobs]
+        ]
+        dest = f"{target.sql_table_id}${partition_id}"
+        # update clustering fields if necessary before copying into place
```

Comment:
This could use more explanation in the code. Am I correct in understanding that this is a performance optimization where we create temp tables without clustering in order to make writes faster, but then have to update the clustering definition in order to satisfy the BQ API's requirement that source and destination tables have the same clustering definition when we `bq cp`?

If that's true, then I worry that our result is going to be essentially unclustered.

Reply:
> Am I correct in understanding that this is a performance optimization where we create temp tables without clustering in order to make writes faster, but then have to update the clustering definition in order to satisfy the BQ API's requirement that source and destination tables have the same clustering definition when we `bq cp`?

yes.

> If that's true, then I worry that our result is going to be essentially unclustered.

I tested it, and the result is effectively clustered.